File: drivers/simulated/control_io.py
from jacobi import Planner
from jacobi.drivers import SimulatedDriver
from jacobi.robots import UniversalUR5e


if __name__ == '__main__':
    robot = UniversalUR5e()
    planner = Planner(robot)

    driver = SimulatedDriver(planner, serve_io_viewer=True, keep_io_viewer_running=True)
    # driver = SimulatedDriver.Controller(keep_io_viewer_running=True)  # Or as standalone controller with I/Os only

    # Create simulated I/O
    driver.create_io('sensor', default_value=1)
    driver.create_io('gripper', default_value=0)

    # Read the value of the digital input
    value = driver.get_digital_input('gripper')
    if value is None:
        print("Could not read from address 'gripper'.")
        exit()

    print(f"Value at digital input 'sensor' is '{value}'.")

    # Set an output for a group of 8 bits
    success = driver.set_digital_output('gripper', 1)
    if not success:
        print("Could not write to address 'gripper'.")
        exit()

    # Read value again
    value = driver.get_digital_input('gripper')
    print(f"Value at digital input 'gripper' is now '{value}'.")

File: drivers/simulated/control_io.cpp
#include <iostream>

#include <jacobi/planner.hpp>
#include <jacobi/drivers/simulated.hpp>
#include <jacobi/robots/universal_ur5e.hpp>


using namespace jacobi;
using namespace jacobi::drivers;
using namespace jacobi::robots;


int main() {
    auto robot = std::make_shared<UniversalUR5e>();
    auto planner = std::make_shared<Planner>(robot);

    auto driver = std::make_shared<SimulatedDriver>(planner, false, false, true, true);
    // auto driver = std::make_shared<SimulatedDriver::Controller>(true, true);  // Or as standalone controller with I/Os only

    // Create simulated I/O
    driver->create_io("gripper", 0);

    // Read the value of the digital input
    auto value = driver->get_digital_input("gripper");
    if (!value) {
        std::cout << "Could not read from address 'gripper'." << std::endl;
        std::exit(1);
    }
    std::cout << "Value of digital input 'gripper' is '" << value.value() << "'." << std::endl;

    // Set an output for a group of 8 bits
    const bool success = driver->set_group_output("gripper", 1);
    if (!success) {
        std::cout << "Could not write to address 'gripper'." << std::endl;
        std::exit(1);
    }

    // Read value again
    value = driver->get_digital_input("gripper");
    std::cout << "Value of digital input 'gripper' is now '" << value.value() << "'." << std::endl;
}